tower-http/src/cors.rs

```diff
@@ -0,0 +1,627 @@
+    /// ```
+    ///
+    /// By default the header will be set to `*`.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+    pub fn allow_origin<T>(mut self, origin: T) -> Self
+    where
+        T: Into<AnyOr<Origin>>,
+    {
+        self.allow_origin = origin.into();
+        self
+    }
+
+    /// Set the value of the [`Access-Control-Expose-Headers`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    /// use http::header::CONTENT_ENCODING;
+    ///
+    /// let layer = CorsLayer::new().expose_headers(vec![CONTENT_ENCODING]);
+    /// ```
+    ///
+    /// All headers can be allowed with
+    ///
+    /// ```
+    /// use tower_http::cors::{CorsLayer, Any};
+    ///
+    /// let layer = CorsLayer::new().expose_headers(Any);
+    /// ```
+    ///
+    /// By default the header will not be set.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
+    pub fn expose_headers<I>(mut self, headers: I) -> Self
+    where
+        I: Into<AnyOr<Vec<HeaderName>>>,
+    {
+        self.expose_headers = Some(match headers.into().0 {
+            AnyOrInner::Any => "*".parse().unwrap(),
+            AnyOrInner::Value(headers) => separated_by_commas(headers),
+        });
+        self
+    }
+}
+
+/// Represents a wildcard value (`*`) used with some CORS headers such as
+/// [`CorsLayer::allow_methods`].
+#[derive(Debug, Clone, Copy)]
+pub struct Any;
+
+/// Used to make methods like [`CorsLayer::allow_methods`] more convenient to call.
+///
+/// You shouldn't have to use this type directly.
+#[derive(Debug, Clone, Copy)]
+pub struct AnyOr<T>(AnyOrInner<T>);
+
+#[derive(Debug, Clone, Copy)]
+enum AnyOrInner<T> {
+    Any,
+    Value(T),
+}
+
+impl From<Origin> for AnyOr<Origin> {
+    fn from(origin: Origin) -> Self {
+        AnyOr(AnyOrInner::Value(origin))
+    }
+}
+
+impl<T> From<Any> for AnyOr<T> {
+    fn from(_: Any) -> Self {
+        AnyOr(AnyOrInner::Any)
+    }
+}
+
+impl<I> From<I> for AnyOr<Vec<Method>>
+where
+    I: IntoIterator<Item = Method>,
+{
+    fn from(methods: I) -> Self {
+        AnyOr(AnyOrInner::Value(methods.into_iter().collect()))
+    }
+}
+
+impl<I> From<I> for AnyOr<Vec<HeaderName>>
+where
+    I: IntoIterator<Item = HeaderName>,
+{
+    fn from(headers: I) -> Self {
+        AnyOr(AnyOrInner::Value(headers.into_iter().collect()))
+    }
+}
+
+fn separated_by_commas<I>(into_iter: I) -> HeaderValue
+where
+    I: IntoIterator,
+    I::Item: ToString,
+{
+    let methods = into_iter
+        .into_iter()
+        .map(|item| item.to_string())
+        .collect::<Vec<_>>()
+        .join(",");
+    HeaderValue::from_str(&methods).unwrap()
+}
+
+impl Default for CorsLayer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<S> Layer<S> for CorsLayer {
+    type Service = Cors<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Cors::new(inner)
```

Comment:
I've tried using the `cors` branch but it appears the CorsLayer options don't actually get used. Could be wrong, but shouldn't this pass the settings down to the Cors struct?

Reply:
Good catch! Fixed in df1d6e001ee6c